### PR TITLE
[prune] Add nncf.prune API

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
### Changes

Add new api `nncf.prune`
Support magnitude based unstructured pruning method 
Example on resnet18

### Related tickets

171577

### Tests

https://github.com/openvinotoolkit/nncf/actions/runs/18034736406

### CODE REVIEW

File: examples/pruning/torch/resnet18/main.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    acc1_init = validate(val_loader, pruned_model, device)
+
+    print(f"Accuracy@1 of pruned model with 70% without fine-tuning: {acc1_init:.3f}")
+
+    ###############################################################################
+    # Step 3: Fine tune with multi step sparsity scheduler
+    print(os.linesep + "[Step 3] Fine tune with multi step sparsity scheduler")
+
+    # Define loss function (criterion) and optimizer.
+    criterion = nn.CrossEntropyLoss().to(device)
```
Comment (daniil-lyakhov) [inline examples/pruning/torch/resnet18/main.py:242]: What's the plan with the RB sparsity loss? Perhaps we can do something like `nncf.SparseLoss` ?

File: examples/pruning/torch/resnet18/main.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    compression_lr = 1e-5
+    optimizer = torch.optim.Adam(pruned_model.parameters(), lr=compression_lr)
+
+    # Create prune scheduler with multi steps strategy
+    prune_scheduler = MultiStepPruneScheduler(
+        pruned_model, mode=PruneMode.UNSTRUCTURED_MAGNITUDE_GLOBAL, steps={0: 0.6, 1: 0.7}
```
Comment (daniil-lyakhov) [inline examples/pruning/torch/resnet18/main.py]: Looks nice and sleek, but I'm not sure how would we use it for the structured pruning algorithm. Looks like compression controller reincarnation in some way or another is unavoidable

File: src/nncf/torch/function_hook/prune/magnitude/algo.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+            # Calculate threshold and binary mask
+            with torch.no_grad():
+                threshold_index = int((data.numel() - 1) * sparsity_level)
+                abs_data = torch.abs(data)
+                threshold = abs_data.view(-1).kthvalue(threshold_index).values
+                new_mask = (abs_data >= threshold).to(dtype=torch.bool)
+
+            # Set new mask
+            hook.binary_mask = new_mask
```
Comment (daniil-lyakhov) [inline src/nncf/torch/function_hook/prune/magnitude/algo.py]: Looks exactly like GLOBAL, but with predefined threshold. Can we do something like this?
```
if mode not in avaliable_modes:
    raise nncf.InternalError
if mode == PruneMode.UNSTRUCTURED_MAGNITUDE_GLOBAL:
   ...
   threshold = ...
for const_name, hook in sparsity_module_map.items():
    ...
```

File: src/nncf/torch/function_hook/prune/prune_model.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+            hook=hook_module,
+        )
+
+    # Set ratio
+    prune_update_ratio(model, mode=mode, ratio=ratio)
+
+    return model
+
+
+def prune_update_ratio(
```
Comment (daniil-lyakhov) [inline src/nncf/torch/function_hook/prune/prune_model.py]: Looks like a duplicate of the function from the `src/nncf/torch/function_hook/prune/magnitude/algo.py`

File: examples/pruning/torch/resnet18/main.py (hunk 1 above)
Comment (AlexanderDokuchaev) [reply examples/pruning/torch/resnet18/main.py:242]: yes it will be extra loss class

File: examples/pruning/torch/resnet18/main.py (hunk 2 above)
Comment (AlexanderDokuchaev) [reply examples/pruning/torch/resnet18/main.py]: it will be we done using scheduler or loss

File: src/nncf/torch/function_hook/prune/magnitude/algo.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    and updates the binary masks accordingly.
+
+    :param model: The neural network model to be pruned.
+    :param mode: The mode of pruning to be applied.
+    :param ratio: The desired pruning ratio, represented as a float between 0 and 1.
+    """
+    pruned_modules_map = get_pruned_modules(model)
+
+    if not pruned_modules_map:
+        msg = "No found Sparsity modules in the model"
```
Comment (ljaljushkin) [inline src/nncf/torch/function_hook/prune/magnitude/algo.py]:
```suggestion
        msg = "No Sparsity modules found in the model"
```

File: src/nncf/torch/function_hook/prune/magnitude/algo.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+def update_ratio(
+    model: nn.Module,
+    mode: PruneMode,
+    ratio: float,
+) -> None:
+    """
+    Updates the pruning ratio for the given model based on the specified pruning mode and sparsity level.
+
+    This function modifies the binary masks of the sparsity modules in the model according to the
+    specified pruning strategy. It calculates he threshold for pruning based on the absolute values of the weights
```
Comment (ljaljushkin) [inline src/nncf/torch/function_hook/prune/magnitude/algo.py]:
```suggestion
    specified pruning strategy. It calculates the threshold for pruning based on the absolute values of the weights
```

File: src/nncf/torch/function_hook/prune/magnitude/algo.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    for unstructured pruning based and update the specified ratio.
+
+    Args:
+        model (nn.Module): The neural network model to be pruned.
+        parameters (list[str]): A list of parameter names to be pruned.
+        mode (PruneMode): The mode of pruning to be applied.
+        ratio (float): The ratio of parameters to prune.
+
+    Returns:
+        nn.Module: The pruned model with hooks registered for the specified parameters.
```
Comment (ljaljushkin) [inline src/nncf/torch/function_hook/prune/magnitude/algo.py]: This one doesn’t match the docstring style we normally use.

File: examples/pruning/torch/resnet18/main.py [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+    print(os.linesep + "[Step 3] Fine tune with multi step sparsity scheduler")
+
+    # Define loss function (criterion) and optimizer.
+    criterion = nn.CrossEntropyLoss().to(device)
+    compression_lr = 1e-5
+    optimizer = torch.optim.Adam(pruned_model.parameters(), lr=compression_lr)
+
+    # Create prune scheduler with multi steps strategy
+    prune_scheduler = MultiStepMagnitudePruningScheduler(
+        pruned_model, mode=PruneMode.UNSTRUCTURED_MAGNITUDE_GLOBAL, steps={0: 0.6, 1: 0.7}
```
Comment (ljaljushkin) [inline examples/pruning/torch/resnet18/main.py]: is it possible to keep a scheduler general and to handle logic for different prunings using mode ```MultiStepPruningScheduler(mode=PruneMode.UNSTRUCTURED_MAGNITUDE_GLOBAL)```
 ?

File: examples/pruning/torch/resnet18/README.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
+# Quantization-Aware Training: An Example for Resnet18 in PyTorch
+
+This example demonstrates how to use Pruning API from Neural Network Compression Framework (NNCF) to prune
+and train PyTorch models on the example of Resnet18 prune aware training, pretrained on Tiny ImageNet-200 dataset.
+
+The example includes the following steps:
+
+- Loading the Tiny ImageNet-200 dataset (~237 Mb) and the Resnet18 PyTorch model pretrained on this dataset.
+- Prune the model using NNCF algorithm by magnitude algorithm.
+- Fine tuning quantized model for one epoch to improve quantized model metrics.
```
Comment (ljaljushkin) [inline examples/pruning/torch/resnet18/README.md]:
```suggestion
- Fine-tuning the pruned model for one epoch to improve its metrics.
```

File: examples/pruning/torch/resnet18/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
+# Quantization-Aware Training: An Example for Resnet18 in PyTorch
+
+This example demonstrates how to use Pruning API from Neural Network Compression Framework (NNCF) to prune
+and train PyTorch models on the example of Resnet18 prune aware training, pretrained on Tiny ImageNet-200 dataset.
```
Comment (ljaljushkin) [inline examples/pruning/torch/resnet18/README.md]:
```suggestion
This example demonstrates how to prune a PyTorch ResNet-18 using NNCF’s Pruning API, then recover accuracy through pruning-aware training, starting with a model initially pretrained on Tiny ImageNet-200.
```

File: examples/pruning/torch/resnet18/README.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
+# Quantization-Aware Training: An Example for Resnet18 in PyTorch
```
Comment (ljaljushkin) [inline examples/pruning/torch/resnet18/README.md]:
```suggestion
# Pruning-Aware Training: An Example for Resnet18 in PyTorch
```

File: src/nncf/torch/function_hook/prune/magnitude/modules.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    :param mask: A binary mask tensor of the same shape as input_ where elements are either 0 or 1.
+
+    :return: The input tensor with the mask applied.
+    """
+    if has_torch_function_unary(input_):
+        return handle_torch_function(apply_magnitude_sparsity_binary_mask, (input_,), input_, mask)  # type: ignore[no-any-return]
+    return input_ * mask
+
+
+class UnstructuredPruneBinaryMask(nn.Module, StatefulModuleInterface):
```
Comment (ljaljushkin) [inline src/nncf/torch/function_hook/prune/magnitude/modules.py]:
```suggestion
class UnstructuredPruningMask(nn.Module, StatefulModuleInterface):
```
what about this name?

File: src/nncf/torch/function_hook/prune/prune_model.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        from nncf.torch.function_hook.prune.magnitude.algo import apply_magnitude_pruning
+
+        model = apply_magnitude_pruning(model, list(parameters_to_sparsity), mode, ratio)
+    else:
+        msg = f"Pruning mode {mode} is not implemented for Torch backend"
+        raise nncf.InternalError(msg)
+    return model
+
+
+def prune_update_ratio(
```
Comment (ljaljushkin) [inline src/nncf/torch/function_hook/prune/prune_model.py]:
```suggestion
def pruning_update_ratio(
```
otherwise it sounds like 2 actions: prune + update

Comment (ljaljushkin) [reply src/nncf/torch/function_hook/prune/prune_model.py]: please also correct in all other places, e.g. `prune_scheduler` -> `pruning_scheduler`

File: src/nncf/torch/function_hook/prune/magnitude/algo.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    # Set ratio
+    update_ratio(model, mode, ratio)
+
+    return model
+
+
+def get_pruned_modules(model: nn.Module) -> dict[str, UnstructuredPruneBinaryMask]:
+    """
+    Retrieves a mapping of operation names to their corresponding
+    MagnitudeSparsityBinaryMask hooks from the given model.
```
Comment (ljaljushkin) [inline src/nncf/torch/function_hook/prune/magnitude/algo.py]: seems like a different name of class

File: examples/pruning/torch/resnet18/main.py (hunk 8 above)
Comment (AlexanderDokuchaev) [reply examples/pruning/torch/resnet18/main.py]: I'm planning rethink it in the next PRs

File: src/nncf/torch/function_hook/prune/prune_model.py (hunk 13 above)
Comment (AlexanderDokuchaev) [reply src/nncf/torch/function_hook/prune/prune_model.py]: Removed `prune_update_ratio`, it does not use anywhere now
Renamed `prune_scheduler`

File: examples/pruning/torch/resnet18/main.py [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+    for epoch in range(2):
+        print(os.linesep + f"Train epoch: {epoch}")
+
+        pruning_scheduler.step()
+
+        train_epoch(train_loader, pruned_model, criterion, optimizer, device=device)
+        acc1 = validate(val_loader, pruned_model, device)
+        # Show statistics of pruning
+        print(
+            f"Accuracy@1 of INT8 model after {epoch} epoch with sparsity "
```
Comment (ljaljushkin) [inline examples/pruning/torch/resnet18/main.py]:
```suggestion
            f"Accuracy@1 of pruned model after {epoch} epoch "
```

File: tests/torch2/function_hook/pruning/magnitude/test_algo.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+    pruned_model = nncf.prune(
+        model, mode=PruneMode.UNSTRUCTURED_MAGNITUDE_LOCAL, ratio=0.5, examples_inputs=example_inputs
+    )
+    pruned_model(example_inputs)
+
+    hook_storage = get_hook_storage(pruned_model)
+    for name, sparsity_module in hook_storage.named_hooks():
+        assert isinstance(sparsity_module, UnstructuredPruningMask)
+        print(name, sparsity_module.binary_mask.device)
```
Comment (ljaljushkin) [inline tests/torch2/function_hook/pruning/magnitude/test_algo.py]:
```suggestion
```

**Review by ljaljushkin — CHANGES_REQUESTED** [review]
minor comments

File: src/nncf/torch/function_hook/prune/magnitude/algo.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+    :param ratio: The ratio of parameters to prune.
+    :returns: The pruned model with hooks registered for the specified parameters.
+    """
+    # Insert hooks
+    pruned_param_names = set()
+    for param_name in parameters:
+        if param_name in pruned_param_names:
+            # To avoid adding multiple hooks to a shared parameters
+            continue
+        pruned_param_names.add(param_name)
```
Comment (daniil-lyakhov) [inline src/nncf/torch/function_hook/prune/magnitude/algo.py]: What is stopping us from doing this?
```suggestion
    for param_name in set(parameters):
```

File: src/nncf/torch/function_hook/prune/magnitude/algo.py [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+
+        cat_all_weights = torch.cat(all_weights).view(-1)
+        threshold_index = int((cat_all_weights.numel() - 1) * ratio)
+        threshold_val = cat_all_weights.kthvalue(threshold_index + 1).values
+
+        for const_name, hook in pruned_modules_map.items():
+            data = get_const_data_by_name(const_name, model)
+
+            # Calculate threshold and binary mask
+            threshold_index = int((data.numel() - 1) * ratio)
```
Comment (daniil-lyakhov) [inline src/nncf/torch/function_hook/prune/magnitude/algo.py]: A redundant line
```suggestion
            # Calculate binary mask
```

File: src/nncf/torch/function_hook/prune/magnitude/schedulers.py [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+    :param model: The neural network model to be pruned.
+    :param mode: The pruning mode to be used.
+    :param steps: A dictionary mapping epochs to sparsity ratios. The keys should be in ascending order,
+                  and the values should be in the range [0, 1).
+    """
+
+    def __init__(self, model: nn.Module, *, mode: PruneMode, steps: dict[int, int]) -> None:
+        super().__init__(model, mode)
+        self.steps = steps
+        if list(self.steps.keys()) != sorted(self.steps.keys()) or any(r >= 1 or r < 0 for r in self.steps.values()):
```
Comment (daniil-lyakhov) [inline src/nncf/torch/function_hook/prune/magnitude/schedulers.py]:
```suggestion
        if list(self.steps) != sorted(self.steps) or any(r >= 1 or r < 0 for r in self.steps.values()):
```

File: src/nncf/torch/function_hook/prune/magnitude/algo.py (hunk 17 above)
Comment (AlexanderDokuchaev) [reply src/nncf/torch/function_hook/prune/magnitude/algo.py]: done

File: src/nncf/torch/function_hook/prune/magnitude/algo.py (hunk 18 above)
Comment (AlexanderDokuchaev) [reply src/nncf/torch/function_hook/prune/magnitude/algo.py]: done

File: src/nncf/torch/function_hook/prune/magnitude/schedulers.py (hunk 19 above)
Comment (AlexanderDokuchaev) [reply src/nncf/torch/function_hook/prune/magnitude/schedulers.py]: done
